pkg/security/podsecuritypolicy/sysctl/mustmatchpatterns.go

```diff
@@ -38,48 +57,89 @@ var (
 
 // NewMustMatchPatterns creates a new mustMatchPatterns strategy that will provide validation.
 // Passing nil means the default pattern, passing an empty list means to disallow all sysctls.
-func NewMustMatchPatterns(patterns []string) SysctlsStrategy {
-	if patterns == nil {
-		patterns = defaultSysctlsPatterns
-	}
+func NewMustMatchPatterns(safeWhitelist, allowedUnsafeSysctls, forbiddenSysctls []string) SysctlsStrategy {
 	return &mustMatchPatterns{
-		patterns: patterns,
+		safeWhitelist:        safeWhitelist,
+		allowedUnsafeSysctls: allowedUnsafeSysctls,
+		forbiddenSysctls:     forbiddenSysctls,
 	}
 }
 
-// Validate ensures that the specified values fall within the range of the strategy.
-func (s *mustMatchPatterns) Validate(pod *api.Pod) field.ErrorList {
-	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, s.validateAnnotation(pod, api.SysctlsPodAnnotationKey)...)
-	allErrs = append(allErrs, s.validateAnnotation(pod, api.UnsafeSysctlsPodAnnotationKey)...)
-	return allErrs
+func (s *mustMatchPatterns) divideSysctls(sysctls []api.Sysctl) (safeSysctls, unsafeSysctls []sysctlItem) {
+	for i, sysctl := range sysctls {
+		found := false
+		for _, ws := range s.safeWhitelist {
+			if sysctl.Name == ws {
+				found = true
+				break
+			}
+		}
+		if found {
+			safeSysctls = append(safeSysctls, sysctlItem{name: sysctl.Name, index: i})
+		} else {
+			unsafeSysctls = append(unsafeSysctls, sysctlItem{name: sysctl.Name, index: i})
+		}
+	}
+	return
 }
 
-func (s *mustMatchPatterns) validateAnnotation(pod *api.Pod, key string) field.ErrorList {
+// Validate ensures that the specified values fall within the range of the strategy.
+func (s *mustMatchPatterns) Validate(pod *api.Pod) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	fieldPath := field.NewPath("pod", "metadata", "annotations").Key(key)
-
-	sysctls, err := helper.SysctlsFromPodAnnotation(pod.Annotations[key])
-	if err != nil {
-		allErrs = append(allErrs, field.Invalid(fieldPath, pod.Annotations[key], err.Error()))
+	var sysctls []api.Sysctl
+	if pod.Spec.SecurityContext != nil {
+		sysctls = pod.Spec.SecurityContext.Sysctls
 	}
 
 	if len(sysctls) > 0 {
-		if len(s.patterns) == 0 {
-			allErrs = append(allErrs, field.Invalid(fieldPath, pod.Annotations[key], "sysctls are not allowed"))
-		} else {
-			for i, sysctl := range sysctls {
-				allErrs = append(allErrs, s.ValidateSysctl(sysctl.Name, fieldPath.Index(i))...)
+		fieldPath := field.NewPath("pod", "spec", "securityContext")
```

Comment:
does the error path really include `pod`?

Reply:
The same path is used in the seccomp validation.